### ISSUE TITLE
A11y: textDirection not set correctly when using attributedXY parameters

### ISSUE THREAD

**Issue by Goddchen** [flutter/flutter#165051]
### Steps to reproduce

We faced an issue that the assert at https://github.com/flutter/flutter/blob/3.29.1/packages/flutter/lib/src/semantics/semantics.dart#L497-L500 was triggering although our widget was placed under a `MaterialApp` and therefore had a text direction set.

Turns out that `_getTextDirection()` just was not checking for `attributedValue`, etc.

### Expected results

`textDirection` is correctly set.

### Actual results

`textDirection` is not set and the assert is being triggered.

### Code sample

```dart
testWidgets('semantics with attributed value should not fail',
    (tester) async {
  await tester.pumpWidget(
    MaterialApp(
      home: Semantics(
        attributedValue: AttributedString('test value'),
        child: const Placeholder(),
      ),
    ),
  );
});
```

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[√] Flutter (Channel stable, 3.24.5, on Microsoft Windows [Version 10.0.26100.3194], locale de-DE)
    • Flutter version 3.24.5 on channel stable at C:\Users\goddc\fvm\versions\3.24.5@stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision dec2ee5c1f (4 months ago), 2024-11-13 11:13:06 -0800
    • Engine revision a18df97ca5
    • Dart version 3.5.4
    • DevTools version 2.37.3

[√] Windows Version (Installed version of Windows is version 10 or higher)

[√] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at C:\Users\goddc\dev\android-sdk
    • Platform android-35, build-tools 34.0.0
    • Java binary at: C:\Program Files\Microsoft\jdk-17.0.13.11-hotspot\bin\java
    • Java version OpenJDK Runtime Environment Microsoft-10376486 (build 17.0.13+11-LTS)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop Windows apps (Visual Studio Community 2022 17.13.2)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.13.35825.156
    • Windows 10 SDK version 10.0.22000.0

[√] Android Studio (version 2024.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 21.0.5+-12932927-b750.29)

[√] IntelliJ IDEA Community Edition (version 2023.1)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2020.1
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart

[√] VS Code, 64-bit edition (version 1.98.1)
    • VS Code at C:\Program Files\Microsoft VS Code
    • Flutter extension version 3.106.0

[√] Connected device (3 available)
    • Windows (desktop) • windows            • windows-x64    • Microsoft Windows [Version
      10.0.26100.3194]
    • Chrome (web)      • chrome             • web-javascript • Google Chrome 134.0.6998.37
    • Edge (web)        • edge               • web-javascript • Microsoft Edge 133.0.3065.69        

[√] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by Goddchen** [comment]
Related PR: https://github.com/flutter/flutter/pull/165028

**Comment by danagbemava-nc** [comment]
Hi @Goddchen, please share the assert that was triggered. The code you linked to points to a different link on the main branch

**Comment by lColinDl** [comment]
https://github.com/flutter/flutter/blob/main/packages/flutter/lib/src/semantics/semantics.dart#L636-L639

**Comment by lColinDl** [comment]
Here's a standalone test to reproduce the issue:
```
testWidgets('semantics with attributed value should not fail',
    (tester) async {
  await tester.pumpWidget(
    MaterialApp(
      home: Semantics(
        attributedValue: AttributedString('test value'),
        child: const Placeholder(),
      ),
    ),
  );
});
```

**Comment by danagbemava-nc** [comment]
Reproducible using the test case provided in https://github.com/flutter/flutter/issues/165051#issuecomment-2720327549


<details>
<summary>test log</summary>

```
══╡ EXCEPTION CAUGHT BY SCHEDULER LIBRARY ╞═════════════════════════════════════════════════════════
The following assertion was thrown during a scheduler callback:
A SemanticsData object with value "test value" had a null textDirection.
'package:flutter/src/semantics/semantics.dart':
Failed assertion: line 670 pos 10: 'attributedValue.string == '' || textDirection != null'

Either the assertion indicates an error in the framework itself, or we should provide substantially
more information in this error message to help you determine and fix the underlying cause.
In either case, please report this assertion by filing a bug on GitHub:
  https://github.com/flutter/flutter/issues/new?template=2_bug.yml

When the exception was thrown, this was the stack:
#2      new SemanticsData (package:flutter/src/semantics/semantics.dart:670:10)
#3      SemanticsNode.getSemanticsData (package:flutter/src/semantics/semantics.dart:3223:12)
#4      SemanticsNode._addToUpdate (package:flutter/src/semantics/semantics.dart:3266:32)
#5      SemanticsOwner.sendSemanticsUpdate (package:flutter/src/semantics/semantics.dart:4031:14)
#6      PipelineOwner.flushSemantics (package:flutter/src/rendering/object.dart:1474:24)
#7      PipelineOwner.flushSemantics (package:flutter/src/rendering/object.dart:1476:15)
#8      AutomatedTestWidgetsFlutterBinding.drawFrame (package:flutter_test/src/binding.dart:1508:35)
#9      RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:495:5)
#10     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1430:15)
#11     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1343:9)
#12     AutomatedTestWidgetsFlutterBinding.pump.<anonymous closure> (package:flutter_test/src/binding.dart:1340:9)
#15     TestAsyncUtils.guard (package:flutter_test/src/test_async_utils.dart:74:41)
#16     AutomatedTestWidgetsFlutterBinding.pump (package:flutter_test/src/binding.dart:1329:27)
#17     WidgetTester.pumpWidget.<anonymous closure> (package:flutter_test/src/widget_tester.dart:599:22)
#20     TestAsyncUtils.guard (package:flutter_test/src/test_async_utils.dart:74:41)
#21     WidgetTester.pumpWidget (package:flutter_test/src/widget_tester.dart:596:27)
#22     main.<anonymous closure> (file:///Users/deanli/projects/QA-Manual/creation/stable_3_29_1/test/widget_test.dart:33:18)
#23     testWidgets.<anonymous closure>.<anonymous closure> (package:flutter_test/src/widget_tester.dart:193:29)
<asynchronous suspension>
#24     TestWidgetsFlutterBinding._runTestBody (package:flutter_test/src/binding.dart:1064:5)
<asynchronous suspension>
<asynchronous suspension>
(elided 7 frames from class _AssertionError, dart:async, and package:stack_trace)
════════════════════════════════════════════════════════════════════════════════════════════════════
```


<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 3.29.2, on macOS 15.3.1 24D70 darwin-arm64, locale en-US) [2.6s]
    • Flutter version 3.29.2 on channel stable at /Users/deanli/dev/stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision c236373904 (10 hours ago), 2025-03-13 16:17:06 -0400
    • Engine revision 18b71d647a
    • Dart version 3.7.2
    • DevTools version 2.42.3
```

```
[!] Flutter (Channel master, 3.30.0-1.0.pre.603, on macOS 15.3.1 24D70 darwin-arm64, locale en-US) [2.9s]
    • Flutter version 3.30.0-1.0.pre.603 on channel master at /Users/deanli/dev/master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision d64db52438 (5 hours ago), 2025-03-13 18:27:29 -0700
    • Engine revision d64db52438
    • Dart version 3.8.0 (build 3.8.0-188.0.dev)
    • DevTools version 2.45.0-dev.0
```
</details>

**Comment by chunhtai** [comment]
agree we should have a check for attributed version

**Comment by Goddchen** [comment]
PR is almost merged, only one of you guys needs to push the button 😉
